File: compiler/bench/lookandsay.c
// expected: 507345
// Look-and-say sequence - describe previous term (15 iterations)
#include <stdio.h>

#define MAXLEN 50000

char buf1[MAXLEN], buf2[MAXLEN];
char *src, *dst;
int slen, dlen;

void init_seq(void) {
    src = buf1; dst = buf2;
    buf1[0] = 1;
    slen = 1;
}

void swap_bufs(void) {
    char *t = src; src = dst; dst = t;
    slen = dlen;
    dlen = 0;
}

void next_seq(void) {
    dlen = 0;
    int i = 0;
    while (i < slen) {
        char ch = src[i];
        int count = 1;
        while (i + count < slen && src[i + count] == ch) {
            count++;
        }
        dst[dlen++] = count + '0';
        dst[dlen++] = ch + '0';
        i += count;
    }
    swap_bufs();
}

int seq_sum(void) {
    int sum = 0;
    for (int i = 0; i < slen; i++) {
        sum += src[i];
    }
    return sum;
}

int main(void) {
    init_seq();
    for (int i = 0; i < 15; i++) {
        next_seq();
    }
    printf("%d\n", seq_sum());
    return 0;
}

File: compiler/bench/lookandsay.fs
\ expected: 507345
\ Look-and-say sequence - describe previous term (15 iterations)

50000 constant MAXLEN
create buf1 MAXLEN allot
create buf2 MAXLEN allot

variable src
variable dst
variable slen
variable dlen

: init-seq ( -- )
  buf1 src !  buf2 dst !
  1 buf1 c!  1 slen ! ;

: swap-bufs ( -- )
  src @ dst @ src ! dst !
  dlen @ slen ! 0 dlen ! ;

: next-seq ( -- )
  0 dlen !
  0                             \ i=0
  begin dup slen @ < while
    dup src @ + c@              \ i char
    1                           \ i char count
    begin
      2 pick 1+ slen @ < if
        2 pick 1+ src @ + c@ 2 pick = if
          1+ swap 1+ swap 1
        else 0 then
      else 0 then
    while repeat
    \ Store count and char
    over 48 + dst @ dlen @ + c!
    dlen @ 1+ dlen !
    48 + dst @ dlen @ + c!
    dlen @ 1+ dlen !
    nip
  repeat drop
  swap-bufs ;

: seq-sum ( -- n )
  0 slen @ 0 do src @ i + c@ + loop ;

: main
  init-seq
  15 0 do next-seq loop
  seq-sum . cr ;
